packages/database/src/migrations/20210813012332-DeleteImmsProjectAndVizes-modifies-data.js

```diff
@@ -0,0 +1,125 @@
+
+const deleteDashboards = async (db, codes) => {
+  const codeList = arrayToDbString(codes);
+
+  await db.runSql(`
+    DELETE FROM dashboard_relation dr
+    USING dashboard d
+    WHERE dr.dashboard_id =  d.id AND d.code IN (${codeList});
+`);
+  await db.runSql(`DELETE FROM dashboard WHERE code in (${codeList})`);
+};
+
+const deleteDashboardItemsByIds = async (db, ids) => {
+  const idList = arrayToDbString(ids);
+
+  await db.runSql(`
+    DELETE FROM dashboard_relation dr
+    USING dashboard_item di
+    WHERE dr.child_id = di.id AND di.id IN (${idList});
+`);
+  await db.runSql(`
+    DELETE FROM legacy_report lr
+    USING dashboard_item di
+    WHERE lr.code = di.code AND di.id IN (${idList});
+`);
+  await db.runSql(`DELETE FROM dashboard_item WHERE id IN (${idList})`);
+};
+
+const deleteMapOverlayGroups = async (db, codes) => {
+  const codeList = arrayToDbString(codes);
+
+  await db.runSql(`
+    DELETE FROM map_overlay_group_relation mogr
+    USING map_overlay_group mog
+    WHERE mogr.map_overlay_group_id = mog.id AND mog.code IN (${codeList});
+  `);
+  await db.runSql(`
+    DELETE FROM map_overlay_group_relation mogr
+    USING map_overlay_group mog
+    WHERE mogr.child_id = mog.id AND mog.code IN (${codeList});
+`);
+  await db.runSql(`DELETE FROM map_overlay_group WHERE code IN (${codeList})`);
+};
+
+const deleteMapOverlaysByIds = async (db, ids) => {
+  const idList = arrayToDbString(ids);
+
+  await db.runSql(`
+    DELETE FROM map_overlay_group_relation mogr
+    USING "mapOverlay" mo
+    WHERE mogr.child_id = mo.id AND mo.id IN (${idList});
+`);
+  await db.runSql(`DELETE FROM "mapOverlay" WHERE id IN (${idList})`);
+};
+
+const deleteProjectAndHierarchy = async (db, code) => {
+  const { rows } = await db.runSql(`SELECT * FROM project WHERE code = '${code}'`);
+  const [{ id: projectId, entity_hierarchy_id: entityHierarchyId }] = rows;
+
+  await db.runSql(`DELETE FROM access_request WHERE project_id = '${projectId}'`);
+  await db.runSql(`DELETE FROM project WHERE code = '${code}'`);
+
+  await toggleEntityTriggers(db, false);
+  await db.runSql(
+    `DELETE FROM ancestor_descendant_relation WHERE entity_hierarchy_id = '${entityHierarchyId}'`,
+  );
+  await db.runSql(`DELETE FROM entity_relation WHERE entity_hierarchy_id = '${entityHierarchyId}'`);
+  await db.runSql(`DELETE FROM entity_hierarchy WHERE id = '${entityHierarchyId}'`);
+  await db.runSql(`DELETE FROM entity WHERE code = '${code}'`);
+  await toggleEntityTriggers(db, true);
+};
+
+const toggleEntityTriggers = async (db, enable) => {
+  const action = enable ? 'ENABLE' : 'DISABLE';
+
+  await db.runSql(`ALTER TABLE entity_relation ${action} TRIGGER entity_relation_trigger`);
+  await db.runSql(`ALTER TABLE entity_hierarchy ${action} TRIGGER entity_hierarchy_trigger`);
+  await db.runSql(`ALTER TABLE entity ${action} TRIGGER entity_trigger`);
+};
+
+exports.up = async function (db) {
+  await deleteDashboards(db, IMMS_DASHBOARDS);
```

Comment:
Same as the map overlays, I think there are some generic dashboard reports that are used in multiple projects, and some of them have `imms` in their project codes. So perhaps we also want to remove `imms` from `project_codes` (dashboard_relation table):

Query I used:

```
select dashboard.* 
from dashboard
join dashboard_relation on dashboard.id = dashboard_relation.dashboard_id
where project_codes::text like '%imms%';
```

Reply:
Great catch!~